File: dongho-kim/2주차/Keys and Rooms/s.py
class Solution:
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        visited = [False] * len(rooms)
        self.dfs(0, rooms, visited)

        for v in visited:
            if not v:
                return False
        return True

    def dfs(self, number, rooms, visited):
        visited[number] = True

        keys = rooms[number]
        for key in keys:
            if visited[key]: continue
            self.dfs(key, rooms, visited)

File: dongho-kim/2주차/Keys and Rooms/s.java
import java.util.*;

class Solution {
    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        boolean[] visited = new boolean[rooms.size()];
        List<Integer> firstRoom = rooms.get(0);
        for (int key : firstRoom) {
            dfs(key, rooms, visited);
        }

        for (int i = 1; i < visited.length; i++) {
            if (!visited[i]) {
                return false;
            }
        }
        return true;
    }

    private void dfs(int num, List<List<Integer>> rooms, boolean[] visited) {
        visited[num] = true;

        List<Integer> nextKeys = rooms.get(num);
        for (int next : nextKeys) {
            if (!visited[next]) {
                dfs(next, rooms, visited);
            }
        }
    }
}
